web-animations-2/Overview.bs

```diff
@@ -969,6 +969,34 @@ Add the following sentence:
 
 </div>
 
+### Calculating the progress of an animation ### {#the-progress-of-an-animation}
+
+  An animation's <dfn export for=animation>progress</dfn> is the ratio of its
+   [=animation/current time=] to its [=associated effect end=].
+
+  <div algorithm="calculate animation progress">
+    The [=animation/progress=] of an [=animation=], |animation|, is calculated as follows:
+
+    <dl class=switch>
+
+    : If <em>any</em> of the following are true:
+      * |animation| does not have an [=animation/associated effect=], or
+      * |animation|'s [=associated effect end=] is zero, or
```

Comment:
Thanks for making these edits. Two follow-up questions that occurred to me:

1. If the associated effect end is zero, should the progress be 1 (or 0 if the current time is negative)? It seems odd that it's unresolved here but maybe I'm just not remembering how all this works?
2. Just to check, should we be clamping the progress to [0, 1]? I assume we shouldn't but I guess it depends on how we expect this property to be used? The answer to this will likely affect how we define the behavior for zero-duration animations.

Reply:
> Thanks for making these edits. Two follow-up questions that occurred to me:
> 
> 1. If the associated effect end is zero, should the progress be 1 (or 0 if the current time is negative)? It seems odd that it's unresolved here but maybe I'm just not remembering how all this works?

I agree, it more accurately reflects the animated progress to return a value here.

> 2. Just to check, should we be clamping the progress to [0, 1]? I assume we shouldn't but I guess it depends on how we expect this property to be used? The answer to this will likely affect how we define the behavior for zero-duration animations.

It seems like it's more useful to allow progresses outside of the [0, 1] range, as especially for 0 it matters to know whether the effect is active or not. That said, allowing values outside of [0, 1] makes it more complicated for how to handle the 0 duration case. Perhaps we should use +/- infinity depending on whether you're before or after the start time?

@bramus @fantasai any thoughts here?

Reply:
Thinking further on this, it may be simpler to restrict to [0, 1] and encourage authors to combine with the phase of the animation which seems like another thing that's not too easy to get today.

Reply:
Yeah, I think you're right. Ultimately, though, it probably depends on how we expect the API to be used.

We _could_ mimic the behaviour of `progress` on `ComputedEffectTiming` and use the fill mode and phase of the root animation effect to make `progress` be `unresolved` when the root effect is not in effect but that might be a bit of a layering violation?

For reference, I believe `ComputedEffectTiming.progress` is mostly in the range [0, 1] but can extend outside that range due to easing functions.

I guess it comes back to reviewing these issues and seeing what would work best in each of those cases (quoting #8799):

> In https://github.com/w3c/csswg-drafts/issues/8669, https://github.com/w3c/csswg-drafts/issues/8765 and https://github.com/w3c/csswg-drafts/issues/8201 there seems to be a common desire to have more convenient APIs for getting some representation of progress.